### PR TITLE
2.x: Switch the connected stream types to track their upstream.

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

File: src/main/java/io/reactivex/flowables/ConnectableFlowable.java [hunk 1]
```diff
@@ -34,8 +34,12 @@
  * @param <T>
  *          the type of items emitted by the {@code ConnectableObservable}
  */
-public abstract class ConnectableFlowable<T> extends Flowable<T> {
-    
+public abstract class ConnectableFlowable<T> extends FlowableWithUpstream<T, T> {
+
+    protected ConnectableFlowable(Publisher<T> source) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/flowables/ConnectableFlowable.java:39]: This mandates `ConnectableFlowable`s to have a source but it is possible that one writes a connectable that emits on its own. Please introduce an intermediate type and leave the Connectables as they are.

**Comment by akarnokd** [comment]
Also there is a compilation error with this.

**Comment by JakeWharton** [comment]
Oops yeah I forgot to push the latest. But I'll fix the hierarchy in a bit.

File: src/main/java/io/reactivex/flowables/ConnectableFlowable.java (hunk 1 above)
Comment (JakeWharton) [inline src/main/java/io/reactivex/flowables/ConnectableFlowable.java:39]: Hmm these are both classes though so there'd be no way to use both in the implementations.

Comment (JakeWharton) [inline src/main/java/io/reactivex/flowables/ConnectableFlowable.java:39]: Unless we make `FlowableFromSource` an interface exposing `source()` and then switch the current class to `AbstractFlowableFromSource`

Comment (akarnokd) [inline src/main/java/io/reactivex/flowables/ConnectableFlowable.java:39]: It will be an interface.

Comment (JakeWharton) [inline src/main/java/io/reactivex/flowables/ConnectableFlowable.java:39]: Ok I'm going to send a new PR with that change for existing operators, and then circle back for these connectable folk.

**Pull request closed without merging**